rustls/src/msgs/codec.rs

```diff
@@ -8,20 +8,20 @@ pub struct Reader<'a> {
 }
 
 impl<'a> Reader<'a> {
-    pub fn init(bytes: &[u8]) -> Reader {
+    pub fn init(bytes: &'a [u8]) -> Reader<'a> {
         Reader {
             buf: bytes,
             offs: 0,
         }
     }
 
-    pub fn rest(&mut self) -> &[u8] {
+    pub fn rest(&mut self) -> &'a [u8] {
```

Comment:
It is. The `&mut self` lifetime is different (shorter) than the `'a` lifetime of the `buf` field.

Reply:
hey @djc, both reader and &[u8] have the same lifetime to bytes which is by default 'a.  the default lifetime works perfectly in this case. 

> This becomes necessary once we start borrowing the reader's contents.
 Not really get the point. reader's contents are bytes shared ref. no matter how many times you borrow, it will not destruct the underlying [u8]. Right ?

Reply:
I remember that some of these changes were necessary in order to more pervasively borrow data throughout rustls. I don't remember exactly which ones. There's an issue here that the `Reader` itself can be substantially shorter-lived than the bytes backing it, and that needed to be expressed in the lifetimes for some of things I wanted to do in #710.